File: Samples/CS/CodeSamples/Server/SMPPServerFAQ.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Net;
using System.Threading.Tasks;
using Inetlab.SMPP;
using Inetlab.SMPP.Common;
using Inetlab.SMPP.PDU;

namespace CodeSamples.Server
{
    public class SMPPServerFAQ
    {
        private readonly SmppServer _server = new SmppServer(new IPEndPoint(IPAddress.Any, 7777));
        private readonly IServerMessageStore _messageStore;

        public SMPPServerFAQ()
        {
            _messageStore = new TestMessageStore();
            _server.evClientSubmitSm += ServerOnClientSubmitSm;
            _server.evClientBind += OnClientBind;
        }

        //<DeliverMessagesOnBind>
        private void OnClientBind(object sender, SmppServerClient client, Bind pdu)
        {
            if (client.BindingMode == ConnectionMode.Transceiver || client.BindingMode == ConnectionMode.Receiver)
            {
                //Start messages delivery

                Task messagesTask = DeliverMessagesAsync(client, pdu);

            }
        }


        private async Task DeliverMessagesAsync(SmppServerClient client, Bind pdu)
        {
            var messages = _messageStore.GetMessagesForClient(pdu.SystemId, pdu.SystemType);

            foreach (TextMessage message in messages)
            {
                var pduBuilder = SMS.ForDeliver()
                    .From(message.PhoneNumber)
                    .To(message.ServiceAddress)
                    .Text(message.Text);

                var responses = await client.Deliver(pduBuilder);

                _messageStore.UpdateMessageState(message.Id, responses);
            }
        }

        public interface IServerMessageStore
        {
            IEnumerable<TextMessage> GetMessagesForClient(string systemId, string systemType);
            void UpdateMessageState(string messageId, DeliverSmResp[] responses);
        }

        public class TextMessage
        {
            public string Id { get; set; }
            public string PhoneNumber { get; set; }
            public string Text { get; set; }

            public string ServiceAddress { get; set; }
        }

        //</DeliverMessagesOnBind>

        //How to send message to connected client
        //
        //<DeliverToClient>
        public async Task DeliverToClient(SmppServerSample.TextMessage message)
        {
            string systemId = GetSystemIdByServiceAddress(message.ServiceAddress);

            SmppServerClient client = FindClient(systemId);

            await client.Deliver(SMS.ForDeliver()
                .From(message.PhoneNumber)
                .To(message.ServiceAddress)
                .Text(message.Text)
            );

        }
        //</DeliverToClient>

        private SmppServerClient FindClient(string systemId)
        {
            return _server.ConnectedClients.FirstOrDefault(c => c.SystemID == systemId);
        }

        private string GetSystemIdByServiceAddress(string serviceAddess)
        {
            if (serviceAddess == "5555") return "ServiceSample";

            return null;
        }

        //How to set MessageId
        // 
        //<SetMessageIdForSubmitSm>
        private void ServerOnClientSubmitSm(object sender, SmppServerClient client, SubmitSm data)
        {
            data.Response.MessageId = Guid.NewGuid().ToString().Substring(0, 8);
        }
        //</SetMessageIdForSubmitSm>


        public class TestMessageStore : IServerMessageStore
        {
            readonly List<TextMessage> _messages = new List<TextMessage>();

            public TestMessageStore()
            {
                _messages.Add(new TextMessage
                {
                    Id = "1",
                    ServiceAddress = "5555",
                    PhoneNumber = "7917123456",
                    Text = "test 1"
                });

                _messages.Add(new TextMessage
                {
                    Id = "2",
                    ServiceAddress = "5555",
                    PhoneNumber = "7917654321",
                    Text = "test 2"
                });
            }

            public IEnumerable<TextMessage> GetMessagesForClient(string systemId, string systemType)
            {
                return _messages;
            }

            public void UpdateMessageState(string messageId, DeliverSmResp[] responses)
            {
                _messages.RemoveAll(m => m.Id == messageId);
            }
        }
    }

    
}

File: Samples/VB.NET/CodeSamples/Server/SMPPServerFAQ.vb
Imports System
Imports System.Collections.Generic
Imports System.Linq
Imports System.Net
Imports System.Threading.Tasks
Imports Inetlab.SMPP
Imports Inetlab.SMPP.Common
Imports Inetlab.SMPP.PDU

Namespace CodeSamples.Server
	Public Class SMPPServerFAQ
		Private ReadOnly _server As New SmppServer(New IPEndPoint(IPAddress.Any, 7777))
		Private ReadOnly _messageStore As IServerMessageStore

		Public Sub New()
			_messageStore = New TestMessageStore()
			AddHandler _server.evClientSubmitSm, AddressOf ServerOnClientSubmitSm
			AddHandler _server.evClientBind, AddressOf OnClientBind
		End Sub

		'<DeliverMessagesOnBind>
		Private Sub OnClientBind(ByVal sender As Object, ByVal client As SmppServerClient, ByVal pdu As Bind)
			If client.BindingMode = ConnectionMode.Transceiver OrElse client.BindingMode = ConnectionMode.Receiver Then
				'Start messages delivery

				Dim messagesTask As Task = DeliverMessagesAsync(client, pdu)

			End If
		End Sub


		Private Async Function DeliverMessagesAsync(ByVal client As SmppServerClient, ByVal pdu As Bind) As Task
			Dim messages = _messageStore.GetMessagesForClient(pdu.SystemId, pdu.SystemType)

			For Each message As TextMessage In messages
				Dim pduBuilder = SMS.ForDeliver().From(message.PhoneNumber).To(message.ServiceAddress).Text(message.Text)

				Dim responses = Await client.Deliver(pduBuilder)

				_messageStore.UpdateMessageState(message.Id, responses)
			Next message
		End Function

		Public Interface IServerMessageStore
			Function GetMessagesForClient(ByVal systemId As String, ByVal systemType As String) As IEnumerable(Of TextMessage)
			Sub UpdateMessageState(ByVal messageId As String, ByVal responses() As DeliverSmResp)
		End Interface

		Public Class TextMessage
			Public Property Id() As String
			Public Property PhoneNumber() As String
			Public Property Text() As String

			Public Property ServiceAddress() As String
		End Class

		'</DeliverMessagesOnBind>

		'How to send message to connected client
		'
		'<DeliverToClient>
		Public Async Function DeliverToClient(ByVal message As SmppServerSample.TextMessage) As Task
			Dim systemId As String = GetSystemIdByServiceAddress(message.ServiceAddress)

			Dim client As SmppServerClient = FindClient(systemId)

			Await client.Deliver(SMS.ForDeliver().From(message.PhoneNumber).To(message.ServiceAddress).Text(message.Text))

		End Function
		'</DeliverToClient>

		Private Function FindClient(ByVal systemId As String) As SmppServerClient
			Return _server.ConnectedClients.FirstOrDefault(Function(c) c.SystemID = systemId)
		End Function

		Private Function GetSystemIdByServiceAddress(ByVal serviceAddess As String) As String
			If serviceAddess = "5555" Then
				Return "ServiceSample"
			End If

			Return Nothing
		End Function

		'How to set MessageId
		' 
		'<SetMessageIdForSubmitSm>
		Private Sub ServerOnClientSubmitSm(ByVal sender As Object, ByVal client As SmppServerClient, ByVal data As SubmitSm)
			data.Response.MessageId = Guid.NewGuid().ToString().Substring(0, 8)
		End Sub
		'</SetMessageIdForSubmitSm>


		Public Class TestMessageStore
			Implements IServerMessageStore

			Private ReadOnly _messages As New List(Of TextMessage)()

			Public Sub New()
				_messages.Add(New TextMessage With {
					.Id = "1",
					.ServiceAddress = "5555",
					.PhoneNumber = "7917123456",
					.Text = "test 1"
				})

				_messages.Add(New TextMessage With {
					.Id = "2",
					.ServiceAddress = "5555",
					.PhoneNumber = "7917654321",
					.Text = "test 2"
				})
			End Sub

			Public Function GetMessagesForClient(ByVal systemId As String, ByVal systemType As String) As IEnumerable(Of TextMessage) Implements IServerMessageStore.GetMessagesForClient
				Return _messages
			End Function

			Public Sub UpdateMessageState(ByVal messageId As String, ByVal responses() As DeliverSmResp) Implements IServerMessageStore.UpdateMessageState
				_messages.RemoveAll(Function(m) m.Id = messageId)
			End Sub
		End Class
	End Class


End Namespace
